File: day2/day2.2.py
from typing import Generator

def get_ids() -> Generator[tuple[int, int], None, None]:
    with open('input.txt', 'r') as input_file:
        for line in input_file:
            for id_range in line.split(','):
                start, end = id_range.split('-')

                yield int(start), int(end)


def all_parts_equal(id_string: str, num_parts: int) -> bool:
    n: int = len(id_string)

    d, r = divmod(n, num_parts)
    if 0 != r:
        return False

    return all(id_string[i:i + d] == id_string[:d] for i in range(d, n, d))


def sum_invalid_range(start: int, end: int):
    total: int = 0

    for product_id in range(start, end + 1):
        id_string: str = str(product_id)
        n: int = len(id_string)

        if any(all_parts_equal(id_string, parts) for parts in range(2, n + 1)):
            total += product_id

    return total


def sum_invalid() -> int:
    total: int = 0

    for start, end in get_ids():
        total += sum_invalid_range(start, end)

    return total


if __name__ == '__main__':
    print(sum_invalid())

File: day2/day2.2.cpp
#include <algorithm>
#include <sstream>
#include <print>
#include <ranges>
#include <vector>
#include <__ranges/iota_view.h>
#include "../common.h"

[[nodiscard]] auto get_ids() -> std::vector<std::pair<long, long>> {
    std::vector<std::pair<long, long>> id_ranges;

    {
        auto f{AOC_INPUT()};
        std::string line;

        while (std::getline(f, line)) {
            std::istringstream line_stream{line};
            std::string token;

            while (std::getline(line_stream, token, ',')) {
                long first{};
                long second{};
                char dash{};
                std::istringstream{token} >> first >> dash >> second;

                id_ranges.emplace_back(first, second);
            }
        }
    }

    return id_ranges;
}

[[nodiscard]] auto all_parts_equal(const std::string& id_string, const int num_parts) -> bool {
    const auto n{id_string.length()};

    if (modulo(static_cast<int>(n), num_parts) != 0) return false;

    const int d{static_cast<int>(n) / num_parts};

    for (auto i{d}; i < n; i += d)
        if (id_string.compare(i, d, id_string, 0, d) != 0)
            return false;

    return true;
}

[[nodiscard]] auto sum_invalid_range(const long start, const long end) -> long {
    long total{};

    for (const auto product_id : std::ranges::views::iota(start, end + 1)) {
        const auto id_string = std::to_string(product_id);
        const auto n = static_cast<int>(id_string.length());

        if (std::ranges::any_of(std::views::iota(2, n + 1), [&](const int parts) {
            return all_parts_equal(id_string, parts);
        }))
            total += product_id;
    }

    return total;
}

[[nodiscard]] auto sum_invalid() -> long {
    long total{};

    for (auto [start, end] : get_ids())
        total += sum_invalid_range(start, end);

    return total;
}

auto main() -> int {
    std::println("{}", sum_invalid());

    return 0;
}
